src/Elf.cpp

```diff
@@ -0,0 +1,989 @@
+        append(output, contents.begin(), contents.end());
+        if (contents.size() < s.get_size()) {
+            append_zeros(output, s.get_size() - contents.size());
+        }
+        append_padding(output, alignment);
+
+        Shdr<T> shdr;
+        shdr.sh_name = strings.get(s.get_name());
+        safe_assign(shdr.sh_type, s.get_type());
+        safe_assign(shdr.sh_flags, s.get_flags());
+        safe_assign(shdr.sh_offset, offset);
+        safe_assign(shdr.sh_addr, offset);
+        safe_assign(shdr.sh_size, s.get_size());
+        safe_assign(shdr.sh_addralign, alignment);
+
+        shdr.sh_link = 0;
+        shdr.sh_info = 0;
+        safe_assign(shdr.sh_entsize, entsize);
+
+        uint16_t shndx = safe_cast<uint16_t>(shdrs.size());
+        section_idxs[&s] = shndx;
+        shdrs.push_back(shdr);
+        return shndx;
+    };
+
+    // And a helper to get the offset we've given a section.
+    auto get_section_offset = [&](const Section &s) -> uint64_t {
+        uint16_t idx = section_idxs[&s];
+        return shdrs[idx].sh_offset;
+    };
+
+    // We need to define the GOT symbol.
+    uint64_t max_got_size = obj.symbols_size() * 2 * sizeof(addr_t);
+    Section got(".got", Section::SHT_PROGBITS);
+    got.set_alignment(4);
+    got.set_size(max_got_size);
+    got.set_flags(Section::SHF_ALLOC);
+    Symbol got_sym("_GLOBAL_OFFSET_TABLE_");
+    got_sym.define(&got, 0, max_got_size);
+    got_sym.set_type(Symbol::STT_OBJECT);
+    got_sym.set_visibility(Symbol::STV_HIDDEN);
+    Symbol dynamic_sym("_DYNAMIC");
+    dynamic_sym.define(&got, 0, 4);
+    dynamic_sym.set_type(Symbol::STT_OBJECT);
+    got.append_contents((addr_t)0);
+    // On some platforms, GOT slots 1 and 2 are also reserved.
+    got.append_contents((addr_t)0);
+    got.append_contents((addr_t)0);
+
+    // Since we can't change the object, start a map of all of the
+    // symbols that we can mutate. If a symbol from the object is a
+    // key in this map, we use the mapped value instead.
+    std::map<const Symbol *, const Symbol *> symbols;
+    symbols[&dynamic_sym] = &dynamic_sym;
+    for (const Symbol &i : obj.symbols()) {
+        if (i.get_name() == "_GLOBAL_OFFSET_TABLE_") {
+            symbols[&i] = &got_sym;
+        } else {
+            symbols[&i] = &i;
+        }
+    }
+
+    // Get a symbol from a relocation, accounting for the symbol map
+    // above.
+    auto get_symbol = [&](const Relocation &r) {
+        const Symbol *sym = r.get_symbol();
+        if (!sym) {
+            return sym;
+        }
+
+        auto i = symbols.find(sym);
+        if (i != symbols.end()) return i->second;
+        return sym;
+    };
+
+    // Check if a relocation needs a PLT entry, which adds some
+    // additional conditions on top of what the linker implementation
+    // wants.
+    auto needs_plt_entry = [&](const Relocation &r) {
+        const Symbol *s = get_symbol(r);
+        if (!s || s->is_defined()) {
```

Comment:
no local symbols need a PLT entry.

symbols that can be internally bound, need not have a PLT entry.